Fix #2669: Split up the admin page controller

BenHenning opened the pull request:
Please see #2669 for detailed information about how the controller was split up. Also, some additional work was done:
- The admin page's `ng-controller` declaration is at the `html` tag rather than a `body` or `div` tag
- I updated the scroll-to-job-output logic to scroll to the top of the job output rather than the bottom of the page (the previous logic did not work so well with large amounts of output)
- I removed all Jinja variables used outside of the controller, which required adding some more variables to the `GLOBALS` array
- Removed the table-style formatting and just used `md-cards` for subsequent directives (they didn't need the table structure anymore, since some of them weren't even really using it)
- Some other minor renames and changes

BenHenning commented:
Let me know if you think it's worthwhile to add the backend API services as part of this refactor or not. I'm okay with these being added as a separate phase of the migration from the initial split up, since it seems like nicely independent work from splitting up controllers.

seanlip commented on core/templates/dev/head/pages/admin/activities_tab/AdminDevModeActivitiesTabDirective.js:
```diff
@@ -0,0 +1,125 @@
+      scope: {
+        setStatusMessage: '='
+      },
+      templateUrl: 'pages/admin/dev_mode_activities_tab_directive',
```
Is this going to be our standard for template urls going forward? If so, might want to add it to the refactor doc and file an issue to change the existing templateUrl names and add a test that scans all the files and ensures that the template url strings match the file names/locations.

seanlip replied:
OK, in that case, maybe just keep the existing format of something like admin/activitiesTab so that this pattern doesn't spread.

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/pages/admin/activities_tab/admin_dev_mode_activities_tab_directive.html:
```diff
@@ -0,0 +1,46 @@
+  <md-card class="oppia-page-card oppia-long-text" style="max-width: 700px">
+    <div class="container" style="margin-left: 0;">
+      <h3>Reload a single exploration</h3>
+      {% for exploration in demo_explorations %}
```
I don't think it's a good idea to have jinja stuff all over the place in random directives. Can we get rid of it?

BenHenning replied:
Agreed; I didn't mean to leave any Jinja in the directives. Removed (here and elsewhere).

seanlip commented on core/templates/dev/head/pages/admin/admin.html:
```diff
@@ -19,406 +26,38 @@
-      {% endif %}
+    <div style="padding-top: 16px;">
+      <admin-dev-mode-activities-tab ng-if="isActivitiesTabOpen() && inDevMode"
+                            set-status-message="setStatusMessage">
```
I think this needs to be indented a bit more; you probably wrote it when the tag name was still admin-activities-tab, I'm guessing :)

BenHenning replied:
Done. Good catch!

seanlip commented on core/templates/dev/head/pages/admin/admin.html:
```diff
@@ -429,10 +68,24 @@
+    <script src="{{TEMPLATE_DIR_PREFIX}}/pages/admin/AdminNavbarDirective.js"></script>
+    <script src="{{TEMPLATE_DIR_PREFIX}}/pages/admin/AdminRouterService.js"></script>
+    <script src="{{TEMPLATE_DIR_PREFIX}}/pages/admin/AdminTaskManagerService.js"></script>
+    <script src="{{TEMPLATE_DIR_PREFIX}}/pages/admin/activities_tab/AdminDevModeActivitiesTabDirective.js"></script>
```
For the activities tab, could we only load the relevant templates and JS files depending on whether it's in dev/prod mode? That way, in prod mode, there is absolutely no chance of a reload comment being sent from this page to the backend.

BenHenning replied:
I don't think restricting it in the frontend is quite as important. I'm going to update the backend controller to actually make it impossible to send the request in production, regardless of what's showing up in the frontend. Also, in terms of performance I don't think Angular will download the HTML for the directive if it doesn't instantiate the directive (this will be more beneficial when we're not including the HTML directly in the directive). There's no way to avoid loading the JS for the directive except via Jinja, and I don't want to proliferate using Jinja anywhere.

seanlip replied:
Sgtm. Thanks!

seanlip commented on core/templates/dev/head/pages/admin/jobs_tab/admin_jobs_tab_directive.html:
```diff
@@ -0,0 +1,124 @@
+        <th>Active realtime layer</th>
+        <th>Controls</th>
+      </tr>
+      {% for computation in continuous_computations_data %}
```
Should get rid of Jinja, I think.

BenHenning replied:
Done.

seanlip commented on core/templates/dev/head/pages/admin/misc_tab/AdminMiscTabDirective.js:
```diff
@@ -0,0 +1,80 @@
+        };
+
+        $scope.downloadTopicSimilaritiesFile = function() {
+          window.location.href = ADMIN_TOPICS_CSV_DOWNLOAD_HANDLER_URL;
```
Could we use $window?

BenHenning replied:
Good idea; done (here and elsewhere).

seanlip commented:
Hi @BenHenning -- this looks great! I had a few comments, but I really like how things are split up.

My main comment is to remove the Jinja stuff, so that it doesn't proliferate further into multiple miscellaneous files -- it's easier to keep track of if it's only on the base pages. Otherwise I think this is excellent.

Thanks!

seanlip commented:
Oh, as far as the backend API stuff is concerned, I'm happy for that to be your call. I just don't think the Jinja stuff should proliferate, but I don't know whether removing it requires doing the backend API stuff or if there's an intermediate step.

BenHenning commented:
Thanks @seanlip, I addressed your review comments. I also updated the admin backend controller to ensure reloading explorations and collections can definitely not happen in the production environment. I changed the jobs tab so that the unfinished/finished jobs sections only show up when there's actually data to show.

seanlip commented on core/templates/dev/head/pages/admin/jobs_tab/AdminJobsTabDirective.js:
```diff
@@ -0,0 +1,118 @@
+            job_type: jobType
+          }).then(function() {
+            $scope.setStatusMessage('Job started successfully.');
+            window.location.reload();
```
May want to use $window; ditto elsewhere in this file.
